File: Exponentiation.cpp
#include <iostream>

using namespace std;

int exponentiation_primitive(int x, int n)
{
	if(n==0)
		return 1;
	return x*exponentiation_primitive(x,n-1);
}

int exponentiation_optimized(int x, int n)
{
	if(n==0)
		return 1;
	if(n%2 == 0)
	{
		int val = exponentiation_optimized(x,n/2);
		return val*val;
	}
	return x*exponentiation_optimized(x,n-1);
}

int main()
{
	int x,n;
	cin>>x>>n;
	int result = exponentiation_primitive(x,n);
	cout<<result<<endl;
	result = exponentiation_optimized(x,n);
	cout<<result<<endl;
	return 0;
}

File: Java/Math/Exponentiation.java
import java.util.Scanner;

public class Exponentiation {
    private static int exponentiationPrimitive(int x, int n)
    {
        if(n == 0)
        {
            return 1;
        }
        return x*exponentiationPrimitive(x, n-1);
    }


    private static int exponentiationOptimized(int x, int n)
    {
        if(n == 0)
        {
            return 1;
        }
        if(n%2 == 0)
        {
            int res = exponentiationOptimized(x, n/2);
            return res*res;
        }
        else
        {
            int res = exponentiationOptimized(x, (n-1)/2);
            return x*res*res;
        }
    }
    
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int x = sc.nextInt();
        int n = sc.nextInt(); 
        System.out.println("Exponentiation Primitive: " + exponentiationPrimitive(x, n));
        System.out.println("Exponentiation Optimized: " + exponentiationOptimized(x, n));
        sc.close();
    }
}
